numpy/_core/src/umath/stringdtype_ufuncs.cpp

```diff
@@ -1047,10 +1047,16 @@ all_strings_promoter(PyObject *NPY_UNUSED(ufunc),
                      PyArray_DTypeMeta *op_dtypes[], PyArray_DTypeMeta *signature[],
                      PyArray_DTypeMeta *new_op_dtypes[])
 {
+    if (op_dtypes[0] != &PyArray_StringDType
+            && op_dtypes[1] != &PyArray_StringDType
+            && op_dtypes[1] != &PyArray_StringDType) {
+        /* No string DType involved after all, do not apply promoter */
+        return 0;
+    }
```

Comment:
I think multiply also needs a similar check, it seems the string promoter is getting called for a multiply reduction on an array of doubles:

```
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGUSR1
    frame #0: 0x0000000186dc5af8 libsystem_kernel.dylib`__kill + 8
libsystem_kernel.dylib`:
->  0x186dc5af8 <+8>:  b.lo   0x186dc5b18               ; <+40>
    0x186dc5afc <+12>: pacibsp
    0x186dc5b00 <+16>: stp    x29, x30, [sp, #-0x10]!
    0x186dc5b04 <+20>: mov    x29, sp
Target 0: (python3.11) stopped.
(lldb) c
Process 67828 resuming
Promoting for 'multiply' promoters only: 0
    DTypes: (None, <class 'numpy.dtypes.Float64DType'>, None)
Promoting for 'multiply' promoters only: 1
    DTypes: (None, <class 'numpy.dtypes.Float64DType'>, None)
Promoting for 'multiply' promoters only: 0
    DTypes: (<class 'numpy.dtypes.StringDType'>, <class 'numpy.dtypes.Float64DType'>, <class 'numpy.dtypes.StringDType'>)
Process 67828 exited with status = 0 (0x00000000)
(lldb) quit

goldbaum at Nathans-MBP in ~/Documents/numpy on main!
± cat ../numpy-experiments/test.py
import os
import signal

import numpy as np

PID = os.getpid()

def do_nothing(*args):
    pass

signal.signal(signal.SIGUSR1, do_nothing)

os.kill(PID, signal.SIGUSR1)

arr = np.array([1.0, 2.0], dtype='float64')
np.multiply.reduce(arr)
```

Reply:
I almost feel like given the existence of the footgun, every single promoter needs a check like this.

Reply:
I wonder if at registering a promotor, it should be possible to pass in a dtype that has to be among the arguments. Might be most useful if that dtype can be abstract (so e.g. any floating kind would do).

Reply:
Yeah, but isn't the matching dynamics just a simple but more powerful way to spell it (if a bit of a boilerplate)?
Unless it would help with disambiguifying things, but then you need a very specific additional rule threaded through everywhere.

If you can clearly formulate it in a way that works with possibly other inhomogeneous ones and doesn't add too much complexity, I would be happy.  It probably could work as a different set of registered loops, but I do we still need this one anyway, probably?

So... I guess its a maybe, but definitely would need a whole second way, although I guess that second way might be enough for promoters relatively genrically.

Reply:
So, it's right that we could in principle have a scheme that differs from the "tuple of dtypes".  Either by having a single DType that needs to be contained or an additional one (although others could be unspecified).

So yes, we can do that, I think it could be used to prioritize only if there is no priority established by other entries unfortunately.  (Maybe at the same level as a specified output dtype.)

So I think either way you end up relying mostly on "matches" logic and it seems not bad to just allow it in the promoter itself.

I don't hate adding it, probably easy enough.  When I did add things first, I had to remove the correct logic for abstract dtypes, because it was unused.

But it'll need to be reviewed also, and we should prioritize gettings things out, even if that means early-adopter DTypes have some limitations around reductions  (let's face it, it was always clear that user-DTypes will only be good after 1-2 releases with them).

Reply:
Actually, thinking a bit more:  There is a better far more direct solution.  We should just never match if an input is `NULL` *unless* that promoter registers with `None` as first DType (or maybe also `np.dtype`) (thus being clear about wanting to handle reductions).

In almost all cases, we'll have an implicit promotion of `(NULL, DType, NULL)` to `(DType, DType, NULL)` which should cover basically any reduction pattern anyway.

That says *nothing* about the merit of the other things brought up, just that it is the most obvious solution for the problem explicitly at hand here.
I do think they have merit, and I think a pattern of `(in1DT, in2DT, outDT, OptionalMustIncludeDType)` may be good, even if I am not sure it can add anything that the `match` return doesn't add, but it probably covers almost all use-cases.